### ISSUE TITLE
MixedLM.predict inconsistently fails for df with string

### ISSUE THREAD

**Issue by kenjioman** [statsmodels/statsmodels#4735]
Reproducible minimum example:
```python
import numpy as np
import pandas as pd
import statsmodels.api as sm

data = pd.DataFrame({'A': np.arange(10),
                     'B': ['a'] * 10,
                     'G': [0]*5 + [1] * 5,
                     'Y': np.arange(10) + np.random.randn(10)})
print(data.head())

model = sm.MixedLM.from_formula('Y ~ A', groups='G', re_formula='1',
                                data=data.iloc[1:,])
result = model.fit()

# This works
print(result.predict(data.head(1)))

# This fails with AttributeError: 'DataFrame' object has no attribute 'dtype'
print(result.predict(data.iloc[0, :]))
```
**If you run it without column "B", it works every time.**

_Also, there are times that we get the_

> ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals

_warning as well, but if you just re-run it a few more times, it'll work (through the fitting part) without any warnings, and the final error will still show up, regardless of the warning._

From `show_versions` (running with Anaconda distribution):
```
INSTALLED VERSIONS
------------------
Python: 3.6.5.final.0
OS: Linux 3.13.0-143-generic #192-Ubuntu SMP Tue Feb 27 10:45:36 UTC 2018 x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

Statsmodels
===========

Installed: 0.9.0 (/path/to/miniconda3/envs/virtual/lib/python3.6/site-packages/statsmodels)

Required Dependencies
=====================

cython: Not installed
numpy: 1.14.3 (/path/to/miniconda3/envs/virtual/lib/python3.6/site-packages/numpy)
scipy: 1.1.0 (/path/to/miniconda3/envs/virtual/lib/python3.6/site-packages/scipy)
pandas: 0.23.0 (/path/to/miniconda3/envs/virtual/lib/python3.6/site-packages/pandas)
    dateutil: 2.7.3 (/path/to/miniconda3/envs/virtual/lib/python3.6/site-packages/dateutil)
patsy: 0.5.0 (/path/to/miniconda3/envs/virtual/lib/python3.6/site-packages/patsy)

Optional Dependencies
=====================

matplotlib: 2.2.2 (/path/to/miniconda3/envs/virtual/lib/python3.6/site-packages/matplotlib)
    backend: module://ipykernel.pylab.backend_inline 
cvxopt: Not installed
joblib: Not installed

Developer Tools
================

IPython: 6.4.0 (/path/to/miniconda3/envs/virtual/lib/python3.6/site-packages/IPython)
    jinja2: 2.10 (/path/to/miniconda3/envs/virtual/lib/python3.6/site-packages/jinja2)
sphinx: 1.7.4 (/path/to/miniconda3/envs/virtual/lib/python3.6/site-packages/sphinx)
    pygments: 2.2.0 (/path/to/miniconda3/envs/virtual/lib/python3.6/site-packages/pygments)
nose: Not installed
pytest: Not installed
virtualenv: Not installed
```

**Comment by kshedden** [comment]
`predict` needs a DataFrame, not a Series as its exog argument.  `data.iloc[0,:]` is a Series.  The following works for me:

```
result.predict(data.iloc[0:1, :])
```

**Comment by kshedden** [comment]
Regarding convergence, depending on the model specification and data, some mixed models are difficult to get to converge.  If you can share model more information we might be able to help.

**Comment by josef-pkt** [comment]
In the example B is perfectly collinear with the constant, but I don't know how MixedLM handles collinear fixed effects (nor what patsy does in this case during categorical conversion).

**Comment by kshedden** [comment]
@josef-pkt in general, is predict supposed to work when `exog.ndim == 1`?  If so, we can add something like `atleast2d` here.  Or we can check `ndim` in predict and raise if `ndim != 2`, so that the behavior is more consistent.

**Comment by josef-pkt** [comment]
@kshedden  Yes, in most predict we have the necessary reshape, but we need to check that model.exog.ndim is not 1.
see `base.model.Results.predict`

numpy arrays are unambiguous to reshape to 2-D, but I think reshaping pandas Series is not unambiguous in all cases when patsy does some transformation on the exog. (I don't remember what I tried for Series.) In those cases users need to disambiguate, but there is no specific exception.

**Comment by josef-pkt** [comment]
To clarify the problem with pandas Series and patsy: 
If exog is a pandas Series and base.model.exog.shape[1] > 1, then it could be that patsy created several exog columns from a single dataframe column. So predict cannot know whether it should be a one row or a one column DataFrame.
AFAIK, there is no (easy) way to get the information from the design_info about how many and which columns in a DataFrame were used in creating the design matrix.

**Comment by kenjioman** [comment]
Just to verify for my specific use case -- the above example was created to try to replicate the error, where my work was with actual data.  In that, I had thought I had accounted for making the exog of the predict function a dataframe, but apparently I had missed something somewhere, as when I try it now, _**it is working fine.  Thanks for your help!**_

That said, yes, I can see the issue with feeding in a series vs dataframe.  As a naive question -- could you not track what features were used in the patsy formula?  As I tried to replicate in my example above, "B" was not even used in the model, and is just an extra column floating with the dataframe (which Patsy, correctly, drops out).  If you know what columns (labels) were used, even with a series input, one could presumably identify what processing needs to be done.

In any case, you've helped me identify my bug/ know that I need to be extra careful in what I feed into the predict function, so my specific needs are met.  Feel free to close this issue, or keep it open (if you want it to track the what to do with Series input issue).

**Comment by kenjioman** [comment]
PS, I remembered what I had done to get back a dataframe.  I had run:
`print(result.predict(data.iloc[0, :].to_frame().T))`, and that was giving me the same AttributeError, even though the input is definitely a dataframe.

**Comment by josef-pkt** [comment]
> As a naive question -- could you not track what features were used in the patsy formula?

No, patsy doesn't give us that information, all indirect ways to get this information would also be tedious and fragile, ie. we would need to scan the formula string ourselves, AFAIK.

**Comment by kenjioman** [comment]
@josef-pkt Ok, thank you once again for the clarification!

**Comment by kshedden** [comment]
Using `data.iloc[0, :].to_frame().T` converts all the columns to `object` dtype.  That seems to throw something off inside of patsy.  I'm not sure if this is something we should handle or not.

**Comment by josef-pkt** [comment]
If there is something that we can clearly detect as wrong, then we can raise an informative exception.
However, we are or `predict` is just the middleman handing off from pandas to patsy without knowing much about the content. e.g. object dtype could be all categoricals or pandas uses it as most general dtype for mixed types.

I most likely did not use many or any categorical/string example cases when I checked the behavior for pandas exog in predict during the latest refactor of predict in base,Results. So, there might still some improvement possibilities.
The main Series example that broke for me was prediction at mean: `predict(exog=df.mean())`, with only numeric columns.
